tests/controllers/test_elasticsearch_controller.py

```diff
@@ -14,143 +14,154 @@
-    assert mock_es_controller._raise_on_exception is False
-    assert mock_es_controller._refresh_period_secs == 10
-    assert mock_es_controller._client is not None
+@pytest.fixture(scope="module", name="create_controller")
+def create_controller_fixture(connection_kwargs) -> typing.Callable[..., ElasticsearchController]:
 
+    def inner_create_controller(*, connection_kwargs=connection_kwargs, refresh_period_secs=10, **controller_kwargs):
+        return ElasticsearchController(connection_kwargs=connection_kwargs,
+                                       refresh_period_secs=refresh_period_secs,
+                                       **controller_kwargs)
 
-@pytest.mark.use_python
-def test_apply_derive_params_func(mock_es_controller):
-    pickled_func_str = str(pickle.dumps(mock_derive_params), encoding="latin1")
+    yield inner_create_controller
 
-    # Create pickled_func_config
-    pickled_func_config = {
-        "pickled_func_str": pickled_func_str,  # Pickled function string
-        "encoding": "latin1"
-    }
 
-    # Apply the mock function and check if connection_kwargs is updated
-    mock_es_controller._apply_derive_params_func(pickled_func_config)
-    assert mock_es_controller._connection_kwargs == {
-        "hosts": [{
-            "host": "localhost", "port": 9200, "scheme": "http"
-        }], "retry_on_status": 3, "retry_on_timeout": 30
-    }
+@pytest.mark.use_python
+def test_constructor(create_controller: typing.Callable[..., ElasticsearchController], connection_kwargs: dict):
+    assert create_controller(raise_on_exception=True)._raise_on_exception is True
+    assert create_controller(refresh_period_secs=1.5)._refresh_period_secs == 1.5
+    assert create_controller()._connection_kwargs == connection_kwargs
 
 
 @pytest.mark.use_python
-def test_refresh_client_force(mock_es_controller):
-    # Simulate a force refresh
-    mock_es_controller.refresh_client(force=True)
+def test_refresh_client_force(create_controller: typing.Callable[..., ElasticsearchController]):
+    controller = create_controller(refresh_period_secs=1)
 
-    assert mock_es_controller._client is not None
-    assert mock_es_controller._last_refresh_time > 0
-    assert isinstance(mock_es_controller._client, Mock)
+    client = controller._client
+    is_refreshed = controller.refresh_client(force=True)
+
+    controller._client.close.assert_called_once()
+    assert client.ping.call_count == 2
+    assert is_refreshed is True
+    assert controller._last_refresh_time > 0
 
 
 @pytest.mark.use_python
-def test_refresh_client_not_needed(mock_es_controller):
+def test_refresh_client_not_needed(create_controller: typing.Callable[..., ElasticsearchController]):
     # Set last refresh time to a recent time
     current_time = time.time()
-    mock_es_controller._last_refresh_time = current_time
+    controller = create_controller()
+    controller._last_refresh_time = current_time
```

Comment:
Suggestion: Do you need to set the value for `_last_refresh_time`? Avoid setting internal members where ever possible

Reply:
Updated as per the suggestions